### ISSUE TITLE
This core needs Cartridge/Disk Slots Implementation. 

### ISSUE THREAD

**Issue by TesterNes** [libretro/blueMSX-libretro#115]
Many people are complaining about SCC cartridge to be used in certain games, or the need for a disk/cassette in conjunction with the game (in cartridge format) to the save the game properly.
In an Msx it's essential to manage what is in Cartridge slot 1 and 2, Disk drive A and B and what Cassette is placed in the cassette player; and a mapper is not enough as it is now. 
These functions are present in the standalone BlueMsx and have to be included in its libreto core. 

It is really a pity because the main issue in standalone Bluemsx is the input lag and thanks to runahead, gpu hard sync and framedelay it's completely gone with Retroarch. But instead we have this big issue.

Please do something about it. Thank you.

**Comment by fry-guy** [comment]
Just chiming in to add my support to this comment. As others have mentioned elsewhere in the issues thread, SD Snatcher is a good example of a rom that won’t run to it’s full potential in RetroArch without the CCS+ cartridge, despite working flawlessly in the standalone PC version of the emulator thanks to the “insert special cartridge” feature.

**Comment by MRIO** [comment]
Indeed, the fact that SCC+ starts in slots 2-3 prevents it working with games such as SD Snatcher, an absolute must play from Hideo Kojima. Project Melancolia translated it from Japanese and I'm aching to play it =]

**Comment by psakhis** [comment]
> Indeed, the fact that SCC+ starts in slots 2-3 prevents it working with games such as SD Snatcher, an absolute must play from Hideo Kojima. Project Melancolia translated it from Japanese and I'm aching to play it =]

You can try configure config.ini with SCCPLUS.ROM attached to slot 2. Works for me.

**Comment by MRIO** [comment]
Thanks Sergi, will try that!

**Comment by MRIO** [comment]
How to actually configure it? I think by config.ini you mean this file:

> ./media/usb0/recalbox/bios/bluemsx.ini
> ./recalbox/share/bios/bluemsx.ini
> ./var/recalboxfs/recalbox/bios/bluemsx.ini

But i do not see how and where to attach the rom. Here’s a sample of the bluemsx.ini where i was searching:

> ports.Eth.ethIndex=-1
> ports.Eth.disabled=0
> ports.Eth.macAddress=00:00:00:00:00:00
> cartridge.defaultType=0
> diskdrive.cdromMethod=none
> diskdrive.cdromDrive=0
> cassette.showCustomFiles=1

Insight appreciated!

**Comment by psakhis** [comment]
Hi, you need SCCPLUS.ROM (on RuMSX emulator for example)

BlueMsx only uses 3 machines, i used /MSXturboR/config.ini.

My config.ini with SCC.ROM (you can try configuring it on standalone bluemsx for example)

```
[Video]
version=V9958
vram size=128kB

[CMOS]
Enable CMOS=1
Battery Backed=1

[Subslotted Slots]
slot 0=1
slot 1=0
slot 2=1
slot 3=1

[External Slots]
slot A=1 0
slot B=2 0


[FDC]
Count=2
[CPU]
Z80 Frequency=3579545Hz
R800 Frequency=7159090Hz
[Board]
type=MSX-T9769C
[Slots]
0 0 0 0 24 "Machines/Shared Roms/KANJI.ROM" ""
0 0 0 0 58 "" ""
0 0 0 0 33 "" ""
0 0 0 0 59 "" ""
0 0 0 0 60 "" ""
0 0 0 0 80 "" ""
0 0 0 0 84 "" ""
0 0 0 0 90 "" ""
0 0 0 0 79 "Machines/Shared Roms/MOONSOUND.rom" ""
0 0 0 0 88 "" ""
0 0 0 4 141 "Machines/Shared Roms/MSXTR.ROM" ""
0 2 2 2 78 "Machines/Shared Roms/MSXTRMUS.ROM" ""
0 3 2 2 42 "Machines/Shared Roms/MSXTROPT.ROM" ""
2 0 2 4 3 "Machines/Shared Roms/SCC.ROM" ""
2 2 2 2 43 "" ""
3 0 0 64 22 "" ""
3 1 0 2 141 "Machines/Shared Roms/MSXTREXT.ROM" ""
3 1 2 4 141 "Machines/Shared Roms/MSXKANJI.ROM" ""
3 2 2 4 143 "Machines/Shared Roms/MSXDOS23.ROM" ""
3 3 2 2 42 "Machines/Shared Roms/XBASIC2.rom" ""

```

**Comment by MRIO** [comment]
A few more pointers that were useful for me in getting the SCC to work:

This page gave me all information i needed to configure the scc card:
http://www.msxblue.com/manual/config_c.htm

And i now know that the config.ini of the machine i needed was in:
/var/recalboxfs/recalbox/bios/Machines/MSX2+

So i thought i would share with anyone coming here! =]

**Comment by Redbeiron** [comment]
> A few more pointers that were useful for me in getting the SCC to work:
> 
> This page gave me all information i needed to configure the scc card: http://www.msxblue.com/manual/config_c.htm
> 
> And i now know that the config.ini of the machine i needed was in: /var/recalboxfs/recalbox/bios/Machines/MSX2+
> 
> So i thought i would share with anyone coming here! =]

can we get display of the config.ini , i have tried doing what is stated above and SD snatcher just says insert sound cart. Also seems like the Cart mapper is broke.

**Comment by MetalGearAlex1** [comment]
Trying to bump this issue to now cause im having same issue. Please FIX! Cart mapper doesnt seem to work even after going into inis

**Comment by thkrz** [comment]
I also can't get it to work, even with the information above.

**Comment by checktext00** [comment]
guys I think I figured it out,

### **short version:**
add
`2 0 2 4 56 "Machines/Shared Roms/SCC.ROM" ""`
to your config.ini and make sure the file "SCC.ROM" exists in the Shared Roms directory (it can be a blank 0 byte file)

this will add the Snatcher SCC-I (aka SCC+) sound cart (cart ID 56) to slot 2.0 and it should work, I've tested it with the SD Snatcher Melancholia translation, the original Snatcher Fudeba translation, and the original Japanese version of Snatcher (Snatcher (1988)(Konami)(JP)[SCC+] - but that one requires the cart to be added to slot 1.0 using `1 0 2 4 56 "Machines/Shared Roms/SCC.ROM" ""` instead)

these are the config.ini files to be changed:

```
RetroArch/system/Machines/MSX2/config.ini
RetroArch/system/Machines/MSX2+/config.ini
RetroArch/system/Machines/MSXturboR/config.ini
```

----
like MRIO said above, https://www.msxblue.com/manual/config_c.htm describes the config.ini setup
(see at the bottom "5th number mapper type" for what I call the cart ID)

the only thing blueMSX cares about is the cart ID, not the contents of the actual SCC.ROM or SCCPLUS.ROM files, so you can just add the correct cart ID and point it to a blank 0 byte file

----

basically there are 3 versions of the SCC cartridge (that I know of) and 3 versions of the SD Snatcher Melancholia translation to fit those 3 carts

- SD Snatcher Sound Cart - SCC-I (aka SCC+) cart originally sold with SD Snatcher (blueMSX cart ID 53)
doesn't work with the "Snatcher" version of the translation

- **Snatcher Sound Cart - SCC-I (aka SCC+) cart originally sold with Snatcher (blueMSX cart ID 56)
WORKS WITH ALL GAMES TESTED**

- generic SCC cart (blueMSX cart ID 3)
only works with the "any SCC" version of the Melancholia translation

----

SD Snatcher Melancholia translation versions ([from the readme file SD092MAN.TXT](https://www.project-melancholia.org/downloads/SD092ENG.ZIP))

- SD Snatcher SCC+ / SD Snatcher Sound Cart version (patch [1] - SD092-1E.IPS)
only works with the Snatcher (cart ID 56) sound cart

- Snatcher SCC+ / Snatcher Sound Cart version (patch [2] - SD092S1E.IPS)
works with the Snatcher (cart ID 56) or SD Snatcher (cart ID 53) cart

- **SCC / SCC Sound Cart version (patch [3] - SD092M1E.IPS)
best version, works with any sound cart, includes the Sound Cart Search Utility to show which carts are detected**
![soundcartutility](https://github.com/user-attachments/assets/5f01e916-f2f3-402d-bc8d-71f5f01bdb50)


hope this helps

**Comment by checktext00** [comment]
I edited the files in the repo and made a pull request #181
